sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp

```diff
@@ -0,0 +1,274 @@
+struct joint_matrix_store_impl<
+    sycl::sub_group, double, 8, 8, Layout, Space,
+    typename std::enable_if_t<Layout == matrix::matrix_layout::row_major ||
+                              Layout == matrix::matrix_layout::col_major>> {
+  void store(
+      matrix::joint_matrix<sycl::sub_group, double,
+                           matrix::matrix_type::accumulator, 8, 8, Layout> &src,
+      multi_ptr<double, Space> dst, size_t stride) {
+
+#ifdef __NVPTX__
+#ifdef __SYCL_DEVICE_ONLY__
+    __dmma_m8n8k4_st_c_f64(dst.get(), src.data, stride,
+                           get_layout_id<Layout>());
+#endif
+#endif
+  }
+};
+
+template <typename Group, typename T1, typename T2, std::size_t M,
+          std::size_t K, std::size_t N, matrix::matrix_layout LayoutA,
+          matrix::matrix_layout LayoutB, matrix::matrix_layout LayoutC,
+          typename Cond = void>
+struct joint_matrix_mad_impl {
+  matrix::joint_matrix<Group, T2, matrix::matrix_type::accumulator, M, N,
+                       LayoutC>
+  mad(Group sg,
+      matrix::joint_matrix<Group, T1, matrix::matrix_type::a, M, K, LayoutA> A,
+      matrix::joint_matrix<Group, T1, matrix::matrix_type::b, K, N, LayoutB> B,
+      matrix::joint_matrix<Group, T2, matrix::matrix_type::accumulator, M, N,
+                           LayoutC>
+          C);
+};
+
+template <matrix::matrix_layout LayoutA, matrix::matrix_layout LayoutB>
+constexpr int get_layout_pair_id();
+
+template <>
+constexpr int get_layout_pair_id<matrix::matrix_layout::row_major,
+                                 matrix::matrix_layout::row_major>() {
+  return 0;
+}
+
+template <>
+constexpr int get_layout_pair_id<matrix::matrix_layout::row_major,
+                                 matrix::matrix_layout::col_major>() {
+  return 1;
+}
+
+template <>
+constexpr int get_layout_pair_id<matrix::matrix_layout::col_major,
+                                 matrix::matrix_layout::row_major>() {
+  return 2;
+}
+
+template <>
+constexpr int get_layout_pair_id<matrix::matrix_layout::col_major,
+                                 matrix::matrix_layout::col_major>() {
+  return 3;
+}
+
+template <matrix::matrix_layout LayoutA, matrix::matrix_layout LayoutB,
+          matrix::matrix_layout LayoutC>
+struct joint_matrix_mad_impl<
+    sycl::sub_group, double, double, 8, 4, 8, LayoutA, LayoutB, LayoutC,
+    typename std::enable_if_t<(LayoutA == matrix::matrix_layout::row_major ||
+                               LayoutA == matrix::matrix_layout::col_major) &&
+                              (LayoutB == matrix::matrix_layout::row_major ||
+                               LayoutB == matrix::matrix_layout::col_major) &&
+                              (LayoutC == matrix::matrix_layout::row_major ||
+                               LayoutC == matrix::matrix_layout::col_major)>> {
+  matrix::joint_matrix<sycl::sub_group, double,
+                       matrix::matrix_type::accumulator, 8, 8, LayoutC>
+  mad(sycl::sub_group sg,
```

Comment:
where is this "mad" defined?
Besides AOT mode, are you planning on supporting JIT mode?

Reply:
"mad" is declared on line 156.  The only current implementation is at this point in the code: now line 200.

I don't think that JIT mode is currently useful for the Tensorcore implementation to use different ptx depending on the compute capability used for execution:  The "wmma" ptx instructions used in the current implementation are hardware generation independent (see the proposal doc for greater detail) and JIT is used to compile the ptx to binary code for each compute capability.
In the future we will add cases that use ptx instructions which are optimized for a particular hardware generation (compute capability).  However at least currently the ptx ISA does not contain multiple instructions which match to a unique case of matrix_type/matrix_use/size that are optimized to different hardware generations - each hardware generation optimized ptx instruction for a given matrix_type uses a different matrix dimension size, and as such will correspond to a unique dpc++ use case.